lib/mapper.js

```diff
@@ -16,41 +18,77 @@ var extend = require('extend');
+  var mappedEvents = [];
+
+  each(function(tag) {
+    if (tag.key === eventName) mappedEvents.push(tag);
+  }, tags);
+
+  // Exit if not mapped
+  if (!mappedEvents.length) return {};
+
   var device = track.device();
+  var properties = track.properties();
   var adTrackingEnabled = device.adTrackingEnabled;
   var options = track.options('DoubleClick Floodlight');
   var copaCompliant = options.copaCompliant;
-  var userAgent = getUserAgent(track);
+  var tags = foldl(function(conversions, tag) {
+    // Required params. These should already be validated at the app level but just in case
+    if (!tag.value.event || !tag.value.cat || !tag.value.type) return conversions;
 
-  // will only send data for mapped events
-  var events = settings.events;
-  var eventName = track.event();
-  var mappedEvent;
+    // Find matching properties if any
+    var matchedVariables = {};
+    each(function(variable) {
+      var segmentProp = variable.key;
+      var floodlightProp = variable.value;
+      var segmentPropValue = properties[segmentProp];
 
-  events.forEach(function(event){
-    if (event.key === eventName){
-      mappedEvent = event.value;
-    }
-  });
+      if (segmentPropValue) matchedVariables[floodlightProp] = segmentPropValue;
+    }, tag.customVariable);
 
-  // Exit if not mapped
-  if (!mappedEvent) return {};
-
-  // Generate random number as shown by Google in their docs
-  // https://support.google.com/dcm/partner/answer/2837459?hl=en&ref_topic=3540316
-  var cacheBuster = Math.random() * 10000000000000000000;
-  var payload = {
-    dc_rdid: device.advertisingId,
-    src: settings.source,
-    cat: mappedEvent.cat,
-    type: mappedEvent.type,
-    ord: cacheBuster,
-    tag_for_child_directed_treatment: copaCompliant ? 1 : 0,
-    dc_lat: adTrackingEnabled ? 1 : 0,
-    userAgent: userAgent
-  };
-
-  return payload;
+    var cacheBuster = Math.random() * 10000000000000000000;
+    var tagParams = {
+      dc_rdid: device.advertisingId,
+      src: settings.source,
+      type: tag.value.type,
+      cat: tag.value.cat,
+      ord: cacheBuster,
+      tag_for_child_directed_treatment: copaCompliant ? 1 : 0,
+      dc_lat: adTrackingEnabled ? 1 : 0,
+      customVariables: matchedVariables
+    };
+
+    conversions.push(tagParams);
+
+    return conversions;
+  }, [], mappedEvents);
+
+  // For each tag, we need to render the GET endpoint with query params
+  var ret = { userAgent: getUserAgent(track), endpoints: [] };
+  each(function(tag) {
+    var endpoint = [
+      'dc_rdid=' + tag.dc_rdid,
+      'src=' + tag.src,
+      'cat=' + tag.cat,
+      'type=' + tag.type,
+      'ord=' + tag.ord,
+      'tag_for_child_directed_treatment=' + tag.tag_for_child_directed_treatment,
+      'dc_lat=' + tag.dc_lat
+    ].join(';');
+
+    each(function(val, key) {
+      // Render each matched custom variable
+      var param = ';' + key + '=' + val
+      endpoint += param;
+    }, tag.customVariables);
```

Comment:
`map` actually just calls `each` and replaces the value with the return value. I had intended a `join` on the end of that, but I think it actually won't work because it still doesn't end up as an array. Ignore this